src/mlpack/tests/ann_layer_test.cpp

```diff
@@ -4078,4 +4078,141 @@ BOOST_AUTO_TEST_CASE(GradientBatchNormWithMiniBatchesTest)
+    {
+      if (i > j)
+        attnMask(i, j) = -INF;
+    }
+  }
+
+  arma::mat keyPaddingMask = arma::zeros(sLen, 1);
+  keyPaddingMask(2, 0) = -INF;
+
+  MultiheadAttention<> module(embedDim, numHeads, dropout, false);
+  module.AttentionMask() = attnMask;
+  module.KeyPaddingMask() = keyPaddingMask;
+  module.Reset();
+
+  //! Forward test.
+  module.Forward(query, output);
+  BOOST_CHECK_EQUAL(output.n_rows, embedDim * tLen);
+  BOOST_CHECK_EQUAL(output.n_cols, bsz);
+
+  //! Backward test.
+  arma::mat gy = 0.01 * arma::randu(embedDim * tLen, bsz);
+  arma::mat g;
+  module.Backward(query, gy, g);
+  BOOST_CHECK_EQUAL(g.n_rows, query.n_rows);
+  BOOST_CHECK_EQUAL(g.n_cols, query.n_cols);
+
+  //! Gradient test.
+  arma::mat error = 0.05 * arma::randu(embedDim * tLen, bsz);
+  arma::mat gradient;
+  module.Gradient(query, error, gradient);
+  BOOST_CHECK_EQUAL(gradient.n_rows, module.Parameters().n_rows);
+  BOOST_CHECK_EQUAL(gradient.n_cols, 1);
+}
+
+/**
+ * Jacobian MultiheadAttention module test.
+ */
+BOOST_AUTO_TEST_CASE(JacobianMultiheadAttentionLayerTest)
+{
+  for (size_t i = 0; i < 5; ++i)
+  {
+    const size_t tgtSeqLen = 1;
+    const size_t srcSeqLen = math::RandInt(2, 10);
+    const size_t embedDim = 8;
+    const size_t nHeads = 2;
+    const size_t batchSize = 1;
+
+    arma::mat input;
+    input.set_size(embedDim * tgtSeqLen, 1);
+
+    MultiheadAttention<> module(embedDim, nHeads);
+    module.Key() = 0.91 * arma::randu(embedDim * srcSeqLen, batchSize);
+    module.Value() = 0.45 * arma::randu(embedDim * srcSeqLen, batchSize);
+
+    double error = JacobianTest(module, input);
+    BOOST_REQUIRE_LE(error, 1e-5);
+  }
+}
+
+/**
+ * Numerical gradient test for MultiheadAttention layer.
+ */
+BOOST_AUTO_TEST_CASE(GradientMultiheadAttentionLayerTest)
+{
+  struct GradientFunction
+  {
+    GradientFunction()
+    {
+      input = arma::randu(embedDim * tgtSeqLen, batchSize);
+      target.set_size(tgtSeqLen, batchSize);
+      for (size_t i = 0; i < target.n_elem; ++i)
+      {
+        target(i) = math::RandInt(0, 2);
+      }
+
+      model = new FFN<MeanSquaredError<>, XavierInitialization>();
+      model->Predictors() = input;
+      model->Responses() = target;
+      model->Add<IdentityLayer<>>();
+      attnModule = new MultiheadAttention<>(embedDim, nHeads, dropout, false);
+      attnModule->Key() = 0.91 * arma::randu(embedDim * srcSeqLen, batchSize);
+      attnModule->Value() = 0.34 * arma::randu(embedDim * srcSeqLen, batchSize);
+      model->Add(attnModule);
+      model->Add<ReLULayer<>>();
+      model->Add<Linear<>>(embedDim * tgtSeqLen, tgtSeqLen);
+      model->Add<Softmax<>>();
+    }
+
+    ~GradientFunction()
+    {
+      delete model;
+    }
+
+    double Gradient(arma::mat& gradient) const
+    {
+      double error = model->Evaluate(model->Parameters(), 0, batchSize);
+      model->Gradient(model->Parameters(), 0, gradient, batchSize);
+      return error;
+    }
```

Comment:
@lozhnikov Can we do something like this here?
```cpp
double Gradient(arma::mat& gradient) const
{
  double error = model->EvaluateWithGradient(model->Parameters(), 0, gradient, batchSize);
  return error;
}
```

Reply:
I think it does the same provided that the `deterministic` equals `true`. Actually `Gradient(4 arguments version)` calls `EvaluateWithGradient()`.